pinc/release_queue.inc

```diff
@@ -20,3 +20,324 @@ function cook_project_selector($project_selector)
+ * @param string $show
+ *   Select which queues to return:
+ *   * all - Don't filter out any queues
+ *   * enabled - Filter out queues which aren't enabled
+ *   * populated - Filter out queues which aren't enabled *OR* contain no projects
+ * @param null|string|array $name_selector
+ *   Select which queues to return in the query where:
+ *   * null - Don't filter out any queues
+ *   * string - Filter out queues whose name does not contain string
+ *   * array - Filter out queues whose name contains none of the strings
+ * @param null|string|array $group_selector
+ *   Select which groups and queues to return:
+ *   * null - Don't filter out any groups/queues.
+ *   * string - Filter out queues/groups whose group name does not contain string
+ *   * array - Filter out queues/groups whose group name contains none of the strings
+ * @param boolean show_groups
+ *   If true, return dummy group objects as well as queues
+ *
+ * @return generator
+ */
+function fetch_queues_data($round, $show, $show_groups, $name_selector = null, $group_selector = null)
+{
+    // Building the WHERE clause is rather ad-hoc, mostly because
+    // a bunch of the output fields don't come directly from the DB result.
+    $proj_where = set_col_str("round_id", $round->id);
+    $clauses = [];
+    if (in_array($show, ["enabled", "populated"])) {
+        $clauses[] = "enabled = 1";
+    }
+    if (is_string($name_selector)) {
+        $name_selector = [$name_selector];
+    }
+    if (is_array($name_selector)) {
+        $values = array_map("DPDatabase::escape", $name_selector);
+        $clauses[] = surround_and_join($values, "name LIKE '%", "%'", ' OR ');
+    }
+    if (empty($clauses)) {
+        $clauses = ["1"];
+    }
+    $filter_where = surround_and_join($clauses, "(", ")", " AND ");
+    // We always want to see the rows that introduce groups.
+    $where = "$proj_where AND ($filter_where OR (SUBSTR(name, 1, 1) = '*'))";
+    $sql = "
+        SELECT *
+        FROM queue_defns
+        WHERE $where
+        ORDER BY ordering
+    ";
+
+    // Because queues are ordered, there is a convention to break them up
+    // into sections with a disabled queue at the start of the section
+    // indicating the section's name. These queue names start with "***"
+    // and we identify them here and treat them differently.
+    $group = null; // the most recently seen row defining a group
+    $res = DPDatabase::query($sql);
+    while ($row = mysqli_fetch_assoc($res)) {
+        if (startswith($row["name"], "***")) {
+            $group = str_replace("***", "", $row["name"]);
+            // Check for group
+            if ($show_groups && str_contains_any_of($group, $group_selector)) {
+                yield [
+                    "is_group" => true,
+                    "group" => $group,
+                ];
+            }
+            continue;
+        }
+
+        if (!is_null($group) && !str_contains_any_of($group, $group_selector)) {
```

Comment:
I think something in here needs tweaking because right now queues without a group (whose return value is `null`) are always included when requesting members of a specific group, which was unexpected.
```bash
curl -i -H "Accept: application/json" -H "X-API-KEY: $API_KEY"
     -X GET "https://www.pgdp.org/~bfoley/c.branch/api-queue/api/v1/queues/rounds/P1?group=General"
```
for instance, includes:
```json
    {
        "id": 14,
        "round_id": "P1",
        "ordering": 540,
        "name": "George Sand Centennary",
        "enabled": true,
        "populated": false,
        "length": 0,
        "unheld_length": 0,
        "comment": "Special queue for Centennary of George Sand",
        "project_selector": "username = 'PM-Sand'",
        "release_criterion": "projects < 10",
        "group": null
    },
```

Reply:
I was playing with this last night -- and queues without a group didn't show up at all; should we add a miscellaneous group for those that don't have one? If I'm not mistaken, everything after the beginning of the first group has to be in a group. So, why doesn't that first set of queues have a group?

Reply:
Yeah, this was the scenario I was trying to handle, but I didn't think through all the corner cases. I think we have 3 cases to worry about (and really cases 2 & 3 are the same).

1. If group is null (ie the query parameter isn't provided), all queues in a group, and all queues with no group should be shown. This is what the `!is_null($group)` workaround was meant to handle, but it's wrong, it should be `!is_null($group_selector)`.
2. If group is set once, then all queues with str_contains($queue_group, $group) should be returned. This will never include queues in no group.
3. If group is set multiple times, then queues with where str_contains($queue_group, $group[i]) for any group will be returned

This does mean that we have no way of explicitly filtering just the 'no group' queues, but, eh... I'm not sure that's a big problem.